### DISCUSSION TITLE
Arm64 Support

### DISCUSSION THREAD

**Discussion by DylanTom** [The-OpenROAD-Project/OpenROAD#2971] [Q&A]
Hi, I am looking to use OpenROAD and have tried to install it on my computer.

It is an M2 Pro Macbook Pro, 32 GB RAM, 1TB SSD. I am running an Ubuntu VM through UTM which has 8GB RAM and 64 GB SSD. 

I am having trouble with building OpenROAD because one of the dependencies is `or-tools`, which does not appear to have Ubuntu support for the ARM architecture. I was wondering if anyone had similar issues or would know how to resolve this.

Specifically, does OpenROAD support the ARM architecture, and if not, is that a priority in the future? 

Thanks!

**Comment by vijayank88** [comment]
Can you try OpenROAD-flow-scripts with docker based: https://openroad-flow-scripts.readthedocs.io/en/latest/user/BuildWithDocker.html#clone-and-build

**Reply by vijayank88** [reply]
@DylanTom you can find `or-tools` releases [here](https://github.com/google/or-tools/releases). It has support for Arm64.

**Reply by DylanTom** [reply]
I'm new to using different architectures and OS. I have the ARM architecture and running Ubuntu 22.04 through a VM. Could you point me to the appropriate package to download?

Thank you!

**Reply by vvbandeira** [reply]
@DylanTom, are you building OR inside a x86 Ubuntu 22.04 VM? If that is the case, the instructions on our docs should be enough ([link here](https://openroad.readthedocs.io/en/latest/main/README.html#install-dependencies)).

Further, if the VM is x86 your problems are likely not related to arm64 architecture. Please provide which commands are you running and the complete output of the command so we can better understand the issue.

**Reply by DylanTom** [reply]
I am currently building OR inside a Linux an ARM64 Ubuntu 22.04 VM. I will try to emulate a x86 Ubuntu 22.04 VM and will reply to this thread.

**Reply by DylanTom** [reply]
On the ARM VM, I ran the following commands
```
git clone --recursive https://github.com/The-OpenROAD-Project/OpenROAD.git
cd OpenROAD
sudo etc/DependencyInstaller.sh
.etc/Build.sh
```
Here I was not able to complete the build with an error that the `-m64` option was not available.

**Reply by DylanTom** [reply]
> @DylanTom, are you building OR inside a x86 Ubuntu 22.04 VM? If that is the case, the instructions on our docs should be enough ([link here](https://openroad.readthedocs.io/en/latest/main/README.html#install-dependencies)).
> 
> Further, if the VM is x86 your problems are likely not related to arm64 architecture. Please provide which commands are you running and the complete output of the command so we can better understand the issue.

Sorry for the late reply, just got a chance to work on this today

I ran the following commands:
```
sudo etc/DependencyInstaller.sh
mkdir build
cd build
cmake .. -DCMAKE_CXX_FLAGS=-march=armv8-a
make
```
On the `make` command, I get the following error message
![image](https://user-images.githubusercontent.com/33362417/224230898-e7aa8e5a-af8f-404b-b991-ccaf8ec0d5f6.png)

**Reply by ZuzooVn** [reply]
@DylanTom same here with Mac M2, Ubuntu 22.04 ARM64 VM. Did you find the solution?

**Reply by DylanTom** [reply]
I did not, will try again in the future. Hopefully, this niche issue would be supported. Would love to use OpenROAD

**Comment by maliberty** [comment]
We do not officially support it as we don't have the hardware resources to test on that platform.  We do try to make sure we can compile on mac M1 based systems.  That said or-tools is available for arm64 as Vijayan pointed out.

**Reply by ZuzooVn** [reply]
@maliberty In OpenROAD tutorial at MICRO 2022, I saw that they can run OpenROAD v2.0-4994 on ARM 64 Lubuntu 20.04 on arm64 Hosts (MacOS)   (https://the-openroad-project.github.io/micro2022tutorial/content/0-prep.html#method-1-install-a-virtual-machine. ) Maybe OpenROAD v2.0-4994 does not include `or-tools`?

**Comment by vvbandeira** [comment]
> Note: OR-Tools only provides archives for the x86_64 (also known as amd64) architecture.
Source: https://developers.google.com/optimization/install/cpp/binary_linux

@DylanTom If you are already using a VM, you can create a x86_64 VM and avoid this issue.

**Reply by ZuzooVn** [reply]
@vvbandeira Since Apple moved to its own processor type (Apple Silicon, i.e. ARM architecture), it's no longer possible to run VMs based on Intel x86-64 on Apple ARM (e.g. on Apple MacBook Pros with M1 or M2 processors).

**Comment by stefanottili** [comment]
If you're on M2, why not give "native" OR a try ?

I've been compiling it on and off using HomeBrew for dependencies (including or-tools) for a while now.
or-tools and mpl2 keep on breaking the build and mpl2 is still be disabled.
>> -- Removing MPL2 and PAR to avoid run time fatal error.

So occasionally it doesn't work, but most of the times I got it to compile in < 15min on a M1.

etc/DependencyInstaller.sh     (unforunately now installs x86 klayout, I prefer building it natively)
./build_openroad.sh --local --latest --clean-force   (breaks today in a mpl2 test)
./build_openroad.sh --local   (when run again it finishes, go figure)

**Reply by DylanTom** [reply]
Hi thanks for responding, could you give a little more detail on how to reproduce what you're describing (i.e. which packages to remove, what packages to install, etc)

Thanks a lot!

**Reply by stefanottili** [reply]
As I said, I used 
etc/DependencyInstaller.sh; ./build_openroad.sh --local
The dependency installer assumes that you have homebrew working.
